### PR TITLE
Better support for D3D12_SHADER_DESC reflection.

### LINKED ISSUE #1540: Hull shader reflection has zero as cControlPoints
I am using IDxcContainerReflection to reflect the DXIL bytecode and get the ID3D12ShaderReflection from it as shown in the Dxc Container Unit Test. But it fails to correctly reflect the number of output control points.
When I compile using fxc and use the regular D3DReflect function, it gets the correct number of output control points.

Code for extracting ID3D12ShaderReflection from DXIL bytecode
```
IDxcLibrary* pLibrary = NULL;
gDxcDllHelper.CreateInstance(CLSID_DxcLibrary, &pLibrary);
IDxcBlobEncoding* pBlob = NULL;
pLibrary->CreateBlobWithEncodingFromPinned((LPBYTE)shaderCode, (UINT32)shaderSize, 0, &pBlob);

IDxcContainerReflection* pReflection;
UINT32 shaderIdx;
gDxcDllHelper.CreateInstance(CLSID_DxcContainerReflection, &pReflection);
pReflection->Load(pBlob);
(pReflection->FindFirstPartKind(DXIL_FOURCC('D', 'X', 'I', 'L'), &shaderIdx));
(pReflection->GetPartReflection(shaderIdx, __uuidof(ID3D12ShaderReflection), (void**)&d3d12reflection));
```
Hull shader file
[grass_tesc.txt](https://github.com/Microsoft/DirectXShaderCompiler/files/2372178/grass_tesc.txt)

The header file generated after compiling the hull shader using dxc. The number of control points is correct here.
[grass.txt](https://github.com/Microsoft/DirectXShaderCompiler/files/2372133/grass.txt)

So it seems that something goes wrong when extracting the ID3D12ShaderReflection object using the container reflection interface

### PR DESCRIPTION
**Pull request by MrTrillian**
A number of properties were unimplemented. I implemented all those that are related to shader properties, not code complexity.

Fixes #1540

### CODE REVIEW

File: lib/HLSL/DxilContainerReflection.cpp [hunk 1]
```diff
@@ -1813,14 +1861,14 @@ HRESULT DxilShaderReflection::GetDesc(D3D12_SHADER_DESC *pDesc) {
...
-  // Unset:  UINT                    GSMaxOutputVertexCount;      // Geometry shader maximum output vertex count
-  // Unset:  D3D_PRIMITIVE           InputPrimitive;              // GS/HS input primitive
-  // Unset:  UINT                    cGSInstanceCount;            // Number of Geometry shader instances
-  // Unset:  UINT                    cControlPoints;              // Number of control points in the HS->DS stage
-  // Unset:  D3D_TESSELLATOR_OUTPUT_PRIMITIVE HSOutputPrimitive;  // Primitive output by the tessellator
-  // Unset:  D3D_TESSELLATOR_PARTITIONING HSPartitioning;         // Partitioning mode of the tessellator
-  // Unset:  D3D_TESSELLATOR_DOMAIN  TessellatorDomain;           // Domain of the tessellator (quad, tri, isoline)
+  pDesc->GSOutputTopology = ToD3DEnum(M.GetStreamPrimitiveTopology());
+  pDesc->GSMaxOutputVertexCount = M.GetMaxVertexCount();
+  pDesc->InputPrimitive = (D3D_PRIMITIVE)M.GetInputPrimitive();
```
Comment (MrTrillian) [inline lib/HLSL/DxilContainerReflection.cpp]: Question for @tex3d : Some code higher up assumes that these enums can be casted, but they are not documented as matching the corresponding D3D enums so I preferred using switch statements. Should I change this one to another `ToD3DEnum` call, or change all of them to straight casts?

```
D3D_PRIMITIVE DxilShaderReflection::GetGSInputPrimitive() {
  if (!m_pDxilModule->GetShaderModel()->IsGS())
    return D3D_PRIMITIVE::D3D10_PRIMITIVE_UNDEFINED;
  return (D3D_PRIMITIVE)m_pDxilModule->GetInputPrimitive();
} 
```

Comment (tex3d) [reply lib/HLSL/DxilContainerReflection.cpp]: A comment should be added to these enums in DxilConstants.h to make it clear that they must match the D3D enums.  They are explicitly set to the same values so they can be cast.  This should allow you to remove translation functions above.

Here are some applicable enums that do not currently have the comment:
DXIL::ShaderKind == D3D11_SHADER_VERSION_TYPE
DXIL::InterpolationMode == D3D_INTERPOLATION_MODE
DXIL::InputPrimitive == D3D_PRIMITIVE
DXIL::PrimitiveTopology == D3D_PRIMITIVE_TOPOLOGY
DXIL::TessellatorDomain == D3D_TESSELLATOR_DOMAIN
DXIL::TessellatorPartitioning == D3D_TESSELLATOR_PARTITIONING
DXIL::TessellatorOutputPrimitive == D3D_TESSELLATOR_OUTPUT_PRIMITIVE

Note: some enums may be lacking values on one side or the other.  This is OK, as some values are used on one side but not the other.  They must agree on the values they both define.

However, for InputPrimitive, I believe you must translate for hull shader case `if (...->IsHS())` using:
`pDesc->InputPrimitive = (D3D_PRIMITIVE)(D3D_PRIMITIVE_1_CONTROL_POINT_PATCH + m_GetInputControlPointCount() - 1);`

File: lib/HLSL/DxilContainerReflection.cpp [hunk 2]
```diff
@@ -1813,14 +1861,14 @@ HRESULT DxilShaderReflection::GetDesc(D3D12_SHADER_DESC *pDesc) {
...
+  pDesc->cGSInstanceCount = M.GetGSInstanceCount();
+  pDesc->cControlPoints = M.GetInputControlPointCount();
```
[the remark this answers is not in the document]
Comment (tex3d) [reply lib/HLSL/DxilContainerReflection.cpp]: For HS, this should be set to GetOutputControlPointCount().  For DS, this should be set to GetInputControlPointCount().  Note that for HS GetInputControlPointCount() gets used to generate the pDesc->InputPrimitive (see other comment).

File: lib/HLSL/DxilContainerReflection.cpp [hunk 3]
```diff
@@ -1785,6 +1785,7 @@ HRESULT DxilShaderReflection::GetDesc(D3D12_SHADER_DESC *pDesc) {
   const ShaderModel *pSM = M.GetShaderModel();
 
   pDesc->Version = EncodeVersion(pSM->GetKind(), pSM->GetMajor(), pSM->GetMinor());
+
   // Unset:  LPCSTR                  Creator;                     // Creator string
```
Comment (MrTrillian) [inline lib/HLSL/DxilContainerReflection.cpp]: Rationale for not filling in more stuff:
- Our current `!llvm.ident` is `clang version 3.7 (tags/RELEASE_370/final)`, and that's probably not useful.
- Converting compilation flags to FXC's format doesn't look all that useful, let's wait to see if anyone depends on this.
- Instruction counts made sense in a world where shader models defined hard limits on them, which I'm not sure is still the case, and in any case they would probably mean something different between FXC and DXC.

**Pull request merged**